src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs

```diff
@@ -182,6 +190,29 @@ internal OpenTelemetryLoggerProvider AddProcessor(BaseProcessor<LogRecord> proce
             return this;
         }
 
+        internal bool ContainsBatchProcessor(BaseProcessor<LogRecord> processor)
+        {
+            if (processor is BatchExportProcessor<LogRecord>)
+            {
+                return true;
+            }
+            else if (processor is CompositeProcessor<LogRecord> compositeProcessor)
+            {
+                var current = compositeProcessor.Head;
+                while (current != null)
+                {
+                    if (this.ContainsBatchProcessor(current.Value))
+                    {
+                        return true;
+                    }
+
+                    current = current.Next;
```

Comment:
I guess this may be improbable but still possible - might have a concurrency issue here in the event another thread changes the value of `Next`.

Reply:
There is the current behavior of `AddProcessor`:
1. If multiple threads are calling `AddProcessor` simultaneously, the caller should be responsible for doing proper synchronization (this follows the .NET general approach - member functions are not thread safe by default).
2. It is safe if a thread is calling `AddProcessor` while other threads are using the processors (e.g. by generating Activity objects which go through the processor chain).

And here is my general thinking regarding thread safety:
1. All the telemetry APIs must be thread safe, following the OpenTelemetry spec:
    * https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#concurrency
    * https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#concurrency-requirements
2. Any other APIs (e.g. the APIs exposed by the SDK package) should follow .NET general guidance, which is:
    * static method should be thread safe
    * member function will not be thread safe
3. Certain functions might provide additional thread safety guarantees in addition to 2), and this should be called out explicitly in the API XML document:
    https://github.com/open-telemetry/opentelemetry-dotnet/blob/9bb7aaf727e64ce738f1113e7b1a10df681bf794/src/OpenTelemetry/BaseExporter.cs#L76-L79

Reply:
👍 makes sense. Agreed we should follow .NET's guidance here.

Different concern, but same area of code... someone could do the following:

```csharp
var processor = new OpenTelemetry.CompositeProcessor<LogRecord>(new SomeProcessor());
loggerProvider.AddProcessor(processor);
processor.AddProcessor(new OpenTelemetry.BatchLogRecordExportProcessor(new SomeExporter()));
```

The provider would end up using use the `ThreadStaticPool`, correct? Maybe this is far fetched as well and not worth doing anything about 😆. I suppose it would require constantly inspecting for any changes to the processor pipeline.

Reply:
Ya that's a problem. Could add an event on `CompositeProcessor` like `internal EventHandler ProcessorAdded` we could fire and use to listen for changes. Just an idea, not going to do anything on this PR to address it.

Reply:
You probably won't like this idea, but could drop the `ThreadStaticPool` for now - if we went with the event solution, definitely seems fit for its own PR. I assume the SharedPool would still have perf benefits even when no batching is involved.

That said, not a blocker at all, I think it's probably fine to roll with this as a known edge case - the thing is, I _think_ it's an edge case. I haven't encountered users who are modifying their processor pipeline after initialization.